### PR TITLE
feat(lsp): provide the deno.cache command server-side

### PR DESCRIPTION
**Pull request by sigmaSd**
fix https://github.com/denoland/deno/issues/20110

### CODE REVIEW

**Comment by bartlomieju** [comment]
@sigmaSd what do you mean by?

> It almost work, but I can't get the referrer

Where do you need the referrer?

File: cli/lsp/language_server.rs [hunk 1]
```diff
@@ -2982,6 +2982,27 @@ impl Inner {
...
+      // We're sending Vec<Vec<uris_to_cache>> in response to deno cache code action
+      let uris: Vec<Vec<String>> = params
+        .arguments
+        .into_iter()
+        .map(|v| serde_json::from_value::<Vec<String>>(v))
+        .map(|v| v.map_err(|e| tower_lsp::jsonrpc::Error::invalid_request())) // FIXME: handle the error
+        .collect::<LspResult<Vec<_>>>()?;
+      return self.cache_request(Some(json!({
+        // TODO: this is wrong, how to get the referrer
+        "referrer": TextDocumentIdentifier{ uri: Url::parse(&uris[0][0]).unwrap()},
```
Comment (sigmaSd) [inline cli/lsp/language_server.rs:2999]: @bartlomieju  The cache request needs the referrer here, as far as I understand it the referrer is the current open document

I'm currently just using the uris which contains the url that the user is going to its definition which is incorrect

The way it works now as far as I can tell:
- uesr hits cache code action
- lsp server responds with the uris to cache
- lsp client responds with the referrer (which is the current active document) and the uris to cache (got it from the server)

maybe the lsp server in step 2 can send the referrer as well (If the info is available) ?

Comment (bartlomieju) [reply cli/lsp/language_server.rs:2999]: @nayeemrmn maybe you could advise here? :)

File: cli/lsp/language_server.rs [hunk 2]
```diff
@@ -2982,6 +2982,27 @@ impl Inner {
...
+        "uris": uris[0].iter().map(|uri|TextDocumentIdentifier{uri: Url::parse(&uri).unwrap()}).collect::<Vec<_>>(),
+      }))).await;
```
Comment (nayeemrmn) [inline cli/lsp/language_server.rs:3001]: Shouldn't the referrer be included in the params? I don't think the LSP can figure out the active document in the editor. Just pass through the argument to be parsed into `CacheParams`.
```suggestion
      return self
        .cache_request(params.arguments.into_iter().next())
        .await;
```

Comment (sigmaSd) [reply cli/lsp/language_server.rs:3001]: The referer is not passed to the arguments, here is an example of helix deno interaction

**a.ts**
```ts
import {} from "https://deno.land/std@0.170.0/testing/bdd.tszzzz";
```

```
2023-09-04T05:39:54.548 helix_lsp::transport [INFO] deno-lsp -> {"jsonrpc":"2.0","method":"textDocument/codeAction","params":{"context":{"diagnostics":[{"code":"no-cache","data":{"specifier":"https://deno.land/std@0.170.0/testing/bdd.tszzzz"},"message":"Uncached or missing remote URL: https://deno.land/std@0.170.0/testing/bdd.tszzzz","range":{"end":{"character":65,"line":0},"start":{"character":15,"line":0}},"severity":1,"source":"deno"}],"triggerKind":1},"range":{"end":{"character":21,"line":0},"start":{"character":16,"line":0}},"textDocument":{"uri":"file:///home/mrcool/dev/deno/lab/tmp/enG/a.ts"}},"id":1}



2023-09-04T05:39:54.593 helix_lsp::transport [INFO] deno-lsp <- {"jsonrpc":"2.0","result":[{"title":"Cache \"https://deno.land/std@0.170.0/testing/bdd.tszzzz\" and its dependencies.","kind":"quickfix","diagnostics":[{"range":{"start":{"line":0,"character":15},"end":{"line":0,"character":65}},"severity":1,"code":"no-cache","source":"deno","message":"Uncached or missing remote URL: https://deno.land/std@0.170.0/testing/bdd.tszzzz","data":{"specifier":"https://deno.land/std@0.170.0/testing/bdd.tszzzz"}}],"command":{"title":"","command":"deno.cache","arguments":[["https://deno.land/std@0.170.0/testing/bdd.tszzzz"]]}},{"title":"Extract function","kind":"refactor.extract.function","isPreferred":false,"disabled":{"reason":"Statement or expression expected."},"data":{"specifier":"file:///home/mrcool/dev/deno/lab/tmp/enG/a.ts","range":{"start":{"line":0,"character":16},"end":{"line":0,"character":21}},"refactorName":"Extract Symbol","actionName":"Extract Function"}},{"title":"Extract constant","kind":"refactor.extract.constant","isPreferred":false,"disabled":{"reason":"Statement or expression expected."},"data":{"specifier":"file:///home/mrcool/dev/deno/lab/tmp/enG/a.ts","range":{"start":{"line":0,"character":16},"end":{"line":0,"character":21}},"refactorName":"Extract Symbol","actionName":"Extract Constant"}},{"title":"Convert default export to named export","kind":"refactor.rewrite.export.named","isPreferred":false,"disabled":{"reason":"Could not find export statement"},"data":{"specifier":"file:///home/mrcool/dev/deno/lab/tmp/enG/a.ts","range":{"start":{"line":0,"character":16},"end":{"line":0,"character":21}},"refactorName":"Convert export","actionName":"Convert default export to named export"}},{"title":"Convert named export to default export","kind":"refactor.rewrite.export.default","isPreferred":false,"disabled":{"reason":"Could not find export statement"},"data":{"specifier":"file:///home/mrcool/dev/deno/lab/tmp/enG/a.ts","range":{"start":{"line":0,"character":16},"end":{"line":0,"character":21}},"refactorName":"Convert export","actionName":"Convert named export to default export"}},{"title":"Convert namespace import to named imports","kind":"refactor.rewrite.import.named","isPreferred":false,"disabled":{"reason":"Selection is not an import declaration."},"data":{"specifier":"file:///home/mrcool/dev/deno/lab/tmp/enG/a.ts","range":{"start":{"line":0,"character":16},"end":{"line":0,"character":21}},"refactorName":"Convert import","actionName":"Convert namespace import to named imports"}}],"id":1}


2023-09-04T05:39:54.593 helix_lsp::transport [INFO] deno-lsp <- [{"command":{"arguments":[["https://deno.land/std@0.170.0/testing/bdd.tszzzz"]],"command":"deno.cache","title":""},"diagnostics":[{"code":"no-cache","data":{"specifier":"https://deno.land/std@0.170.0/testing/bdd.tszzzz"},"message":"Uncached or missing remote URL: https://deno.land/std@0.170.0/testing/bdd.tszzzz","range":{"end":{"character":65,"line":0},"start":{"character":15,"line":0}},"severity":1,"source":"deno"}],"kind":"quickfix","title":"Cache \"https://deno.land/std@0.170.0/testing/bdd.tszzzz\" and its dependencies."},{"data":{"actionName":"Extract Function","range":{"end":{"character":21,"line":0},"start":{"character":16,"line":0}},"refactorName":"Extract Symbol","specifier":"file:///home/mrcool/dev/deno/lab/tmp/enG/a.ts"},"disabled":{"reason":"Statement or expression expected."},"isPreferred":false,"kind":"refactor.extract.function","title":"Extract function"},{"data":{"actionName":"Extract Constant","range":{"end":{"character":21,"line":0},"start":{"character":16,"line":0}},"refactorName":"Extract Symbol","specifier":"file:///home/mrcool/dev/deno/lab/tmp/enG/a.ts"},"disabled":{"reason":"Statement or expression expected."},"isPreferred":false,"kind":"refactor.extract.constant","title":"Extract constant"},{"data":{"actionName":"Convert default export to named export","range":{"end":{"character":21,"line":0},"start":{"character":16,"line":0}},"refactorName":"Convert export","specifier":"file:///home/mrcool/dev/deno/lab/tmp/enG/a.ts"},"disabled":{"reason":"Could not find export statement"},"isPreferred":false,"kind":"refactor.rewrite.export.named","title":"Convert default export to named export"},{"data":{"actionName":"Convert named export to default export","range":{"end":{"character":21,"line":0},"start":{"character":16,"line":0}},"refactorName":"Convert export","specifier":"file:///home/mrcool/dev/deno/lab/tmp/enG/a.ts"},"disabled":{"reason":"Could not find export statement"},"isPreferred":false,"kind":"refactor.rewrite.export.default","title":"Convert named export to default export"},{"data":{"actionName":"Convert namespace import to named imports","range":{"end":{"character":21,"line":0},"start":{"character":16,"line":0}},"refactorName":"Convert import","specifier":"file:///home/mrcool/dev/deno/lab/tmp/enG/a.ts"},"disabled":{"reason":"Selection is not an import declaration."},"isPreferred":false,"kind":"refactor.rewrite.import.named","title":"Convert namespace import to named imports"}]
```

Though I now noticed that helix does send the referrer inside `params.textDocument.uri` I wonder If I can get that

Comment (nayeemrmn) [reply cli/lsp/language_server.rs:3001]: > The referer is not passed to the arguments

I'm not sure I understand, wouldn't the Deno helix extension sending the `deno.cache` command be conscious of the protocol and send the referrer?

Comment (sigmaSd) [reply cli/lsp/language_server.rs:3001]: Helix doesn't have extensions support, the point of this exercise is to try to make cache command work without extensions, by using only lsp spec defined commands in this case I feel like `workspace/executeCommand` almost works https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#workspace_executeCommand

According to the spec, the arugments of `ExecuteCommandParams` are sent from the server to the clinet and then the client resend them back (which is what helix does)

>The arguments are typically specified when a command is returned from the server to the client. Example requests that return a command are [textDocument/codeAction](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_codeAction) or [textDocument/codeLens](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_codeLens).

I think there are 2 ways to handle this:
- deno lsp server sends the referrer with the url to cache, but I think from what I understood of you is that the lsp server is not aware of the active document 
- Maybe use `params.textDocument.uri` from the client codeAction request https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_codeAction (its in the codeAction params), As you can see from the first message of helix it does send the referrer there, the descripton from the spec says `The document in which the command was invoked.` which seems exactly what we want

I think the second method is reasonable, does that seems correct to you ?

Comment (nayeemrmn) [reply cli/lsp/language_server.rs:3050]: > According to the spec, the arugments of `ExecuteCommandParams` are sent from the server to the clinet and then the client resend them back (which is what helix does)

I think I see, so which part of Deno in this repo sends those arguments to the client and when does that occur?

Comment (sigmaSd) [reply cli/lsp/language_server.rs:3050]: I think its when deno responds with to the codeAction so here https://github.com/denoland/deno/blob/8d598dada6d9a7152ba4e537a1789c4cac30e425/cli/lsp/language_server.rs#L1852

But I'll have to debug to be sure

Comment (nayeemrmn) [reply cli/lsp/language_server.rs:3050]: Got it, so just change the arguments passed here: https://github.com/denoland/deno/blob/9e243d22f4ea9642e24415e5484f0f067f466ef5/cli/lsp/diagnostics.rs#L999
to:
```rust
              arguments: Some(vec![serde_json::to_value(
                super::lsp_custom::CacheParams {
                  referrer: lsp::TextDocumentIdentifier {
                    uri: specifier.clone(),
                  },
                  uris: vec![lsp::TextDocumentIdentifier {
                    uri: data.specifier,
                  }],
                },
              )
              .unwrap()]),
```
I'm not sure if we have to change anything in vscode_deno, I'll investigate.

Comment (sigmaSd) [reply cli/lsp/language_server.rs:3050]: perfect works

Comment (nayeemrmn) [reply cli/lsp/language_server.rs:3050]: Actually we have to change it in a backwards compatible way so new Deno doesn't break with old extension, so more like:
```rust
              arguments: Some(vec![json!([data.specifier]), json!(&specifier)]),
```

Comment (sigmaSd) [reply cli/lsp/language_server.rs]: done

**Comment by sigmaSd** [comment]
It would be great if someone add the test, I know what to do but it will take me a while to figure out those json messages.

File: cli/lsp/language_server.rs [hunk 3]
```diff
@@ -3031,6 +3031,36 @@ impl Inner {
...
+    if params.command == "deno.cache" {
+      let mut arguments = params.arguments.into_iter();
+      let uris: Vec<Url> = arguments
+        .next()
+        .and_then(|value| serde_json::from_value(value).ok())
+        .ok_or(LspError::invalid_params("missing or incorrect uris"))?;
+      let referrer: Url = arguments
+        .next()
+        .and_then(|value| serde_json::from_value(value).ok())
+        .ok_or(LspError::invalid_params("missing or incorrect referrer"))?;
```
Comment (nayeemrmn) [inline cli/lsp/language_server.rs:3047]: This lets us maximally defer error messages to serde:
```suggestion
      let uris = serde_json::to_value(arguments.next()).unwrap();
      let uris: Vec<Url> = serde_json::from_value(uris)
        .map_err(|err| LspError::invalid_params(err.to_string()))?;
      let referrer = serde_json::to_value(arguments.next()).unwrap();
      let referrer: Url = serde_json::from_value(referrer)
        .map_err(|err| LspError::invalid_params(err.to_string()))?;
```

**Comment by nayeemrmn** [comment]
> It would be great if someone add the test, I know what to do but it will take me a while to figure out those json messages.

No problem, I can take over the rest of this PR and coordinating it with vscode_deno. There are also some other `deno.cache` command constructions that need to be fixed here.

**Comment by sigmaSd** [comment]
That would be great, I'll leave it to you then thanks.

**Comment by nayeemrmn** [comment]
Looking again it'll be more complicated to not break current and older versions of the extension. The command registration at https://github.com/denoland/deno/pull/20111/files#diff-36f322c6b36e52165bbafb69fd8f4ae78d5f7e36deebc6c8d9d30e62759651b1R122 will conflict with the command registration in the extension.

We'll probably have to do a check on both sides but perhaps they make sense on their own. I'll patch the extension to check that a command isn't already registered before it attempts to register one (probably should anyway), and send `"enableBuiltinCommands": true` in the initialization options. The LSP won't register this command until it gets that option.

**Comment by sigmaSd** [comment]
That will still require an extension to helix to send the custom init message, that's unfortunate but I guess it is what it is

**Comment by nayeemrmn** [comment]
> That will still require an extension to helix to send the custom init message, that's unfortunate but I guess it is what it is

Oh yeah that's unfortunate. I expect in a couple versions we'll remove that option and make them always enabled. I just don't want users to see extension crashes on the same version cycle we're supposed to be making a bunch of improvements to it.

**Comment by sigmaSd** [comment]
You're setting the capability to an empty vec instead of none, this means helix should still work because they only check if the capability is none

**Comment by nayeemrmn** [comment]
> You're setting the capability to an empty vec instead of none, this means helix should still work because they only check if the capability is none

Interesting, though isn't that might be a bug on Helix's end? https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#executeCommandOptions.

**Comment by sigmaSd** [comment]
I agree helix shouldn't execute commands not specified by the server capability. This is probably just a missing feature and it won't be added until someone complain which I doubt it would happen any time soon realistically

**Review by bartlomieju — APPROVED** [review]
LGTM

**Pull request merged**